File: shortest_distance_3.py
# Similar to part 1 - words are duplicate and we want shortest distance
# Time - O(N)
# Space - O(1)

class Solution:
    def shortestWordDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:

        idx1 = -1
        idx2 = -1
        result = float("inf")
        
        for i in range(len(wordsDict)):
            
            if wordsDict[i] == word1:
                idx1 = i
                
            if wordsDict[i] == word2:
                if idx1 == i:
                    idx1 = idx2
                    
                idx2 = i 
            
            if idx1 != -1 and idx2 !=-1:
                
                result = min(result, abs(idx1-idx2))
                
        return result
                
        
        
        
        
       

File: ShortestDistance3.java
//time complexity: O(n)
//space complexity: O(1)
class Solution {
    public int shortestWordDistance(String[] wordsDict, String word1, String word2) {
        int p1=-1;
        int p2=-1;
        int max=Integer.MAX_VALUE;
        if(word1.equals(word2))
        {
            for(int i=0;i<wordsDict.length;i++)
            {
                if(wordsDict[i].equals(word1))
                {
                    if(p1>p2)
                        p2=i;
                    else
                        p1=i;
                }
                if(p1!=-1 && p2!=-1)
                    max=Math.min(max,Math.abs(p2-p1));
            }
        }
        else
        {
            for(int i=0;i<wordsDict.length;i++)
            {
                if(wordsDict[i].equals(word1)){
                    p1=i;
                }
                if(wordsDict[i].equals(word2)){
                    p2=i;
                }
                if(p1!=-1 && p2!=-1)
                    max=Math.min(Math.abs(p2-p1),max);
            }
        }
        return max;
    }
}